src/repo.rs

```diff
@@ -753,42 +795,89 @@ impl DocumentInfo {
-
-                // TODO: remove remote if there is an error.
+                let conn = match self.peer_connections.entry(repo_id.clone()) {
+                    Entry::Vacant(entry) => {
+                        // if this is a new peer, request authorization
+                        commands.push(PeerConnCommand::RequestAuth(repo_id.clone()));
+                        entry.insert(PeerConnection::pending())
+                    }
+                    Entry::Occupied(entry) => entry.into_mut(),
+                };
                 for message in messages {
-                    document
-                        .automerge
-                        .receive_sync_message(sync_state, message)
-                        .expect("Failed to apply sync message.");
+                    conn.receive_sync_message(&mut document.automerge, message)
+                        .expect("Failed to receive sync message.");
                 }
             }
             let new_heads = document.automerge.get_heads();
             (start_heads, new_heads)
         };
-        start_heads != new_heads
+        (start_heads != new_heads, commands)
+    }
+
+    /// Promote a peer awaiting authorization to a full peer
+    ///
+    /// Returns any messages which the peer sent while we were waiting for authorization
+    fn promote_pending_peer(&mut self, repo_id: &RepoId) -> Option<Vec<SyncMessage>> {
+        if let Some(PeerConnection::PendingAuth { received_messages }) =
+            self.peer_connections.remove(repo_id)
+        {
+            self.peer_connections
+                .insert(repo_id.clone(), PeerConnection::Accepted(SyncState::new()));
+            Some(received_messages)
+        } else {
+            tracing::warn!(remote=%repo_id, "Tried to promote a peer which was not pending authorization");
+            None
+        }
     }
 
     /// Potentially generate an outgoing sync message.
     fn generate_first_sync_message(&mut self, repo_id: RepoId) -> Option<SyncMessage> {
-        let sync_state = self.sync_states.entry(repo_id).or_default();
-        let document = self.document.read();
-        document.automerge.generate_sync_message(sync_state)
+        match self.peer_connections.entry(repo_id) {
+            Entry::Vacant(entry) => {
+                let mut sync_state = SyncState::new();
+                let document = self.document.read();
+                let message = document.automerge.generate_sync_message(&mut sync_state);
+                entry.insert(PeerConnection::Accepted(sync_state));
+                message
+            }
+            Entry::Occupied(mut entry) => match entry.get_mut() {
+                PeerConnection::PendingAuth { .. } => {
+                    let mut sync_state = SyncState::new();
+                    let document = self.document.read();
+                    let message = document.automerge.generate_sync_message(&mut sync_state);
```

Comment:
first apply received sync messages? 

So this branch is hit if the peer sends us a sync message while our `announce` request is pending, in a way it happens when two peers connect, and one of them get permission to announce, and we receive the sync message before the local request to announce has completed?